### ISSUE TITLE
PERF403 On Nested For Loops With Body

### ISSUE THREAD

**Issue by amirsoroush** [astral-sh/ruff#22848]
### Summary

Hi Ruff team,

Here is my sample code:

```python
from itertools import batched
import sqlalchemy as sa
from sqlalchemy.ext.asyncio import AsyncSession

class User:
    pass

async def foo(session: AsyncSession, args: set[str]) -> dict[str, int]:
    d: dict[str, int] = {}

    for batch in batched(args, 1000):
        query = sa.select(User.x, User.y).where(User.id.in_(batch))
        results = (await session.execute(query)).tuples().all()
        print("hi")
        print("hi")

        for x, y in results:
            d[y] = x
    return d
```
running ruff with: `uv run ruff check --select PERF403 filename.py` gives:

```none
PERF403 Use a dictionary comprehension instead of a for-loop
  --> filename.py:18:13
   |
17 |         for x, y in results:
18 |             d[y] = x
   |             ^^^^^^^^
19 |     return d
   |
help: Replace for loop with dict comprehension

Found 1 error.
```

Preferably I'd like to see no error in the line `d[y] = x`. The body can be quite complex and it's not a direct nested for loop. The inner loop fills a dictionary(`d`) which is defined outside of the outer loop.

Thanks

### Version

0.14.14

**Comment by ntBre** [comment]
Thanks for the report!

Just to clarify, this is the suggested fix for the diagnostic:

```diff
         print("hi")
         print("hi")

-        for x, y in results:
-            d[y] = x
+        d.update({y: x for x, y in results})
     return d
```

So I think the rule is already taking into account the complexity of the outer loop.

If you'd still prefer to avoid the `update` call in that case, I think this is a duplicate of #21890.
